src/python/pants/core/goals/publish.py

```diff
@@ -0,0 +1,226 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from abc import ABCMeta
+from dataclasses import dataclass
+from itertools import chain
+from typing import ClassVar, Generic, Iterable, Type, TypeVar
+
+from typing_extensions import final
+
+from pants.core.goals.package import BuiltPackage, PackageFieldSet
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.process import InteractiveProcess, InteractiveRunner
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule, rule
+from pants.engine.target import (
+    FieldSet,
+    NoApplicableTargetsBehavior,
+    TargetRootsToFieldSets,
+    TargetRootsToFieldSetsRequest,
+)
+from pants.engine.unions import UnionRule, union
+
+logger = logging.getLogger(__name__)
+
+
+_F = TypeVar("_F", bound=FieldSet)
+
+
+@union
+@dataclass(frozen=True)
+class PublishPackagesRequest(Generic[_F]):
+    field_set: _F
+    packages: tuple[BuiltPackage, ...]
+
+
+_T = TypeVar("_T", bound=PublishPackagesRequest)
+
+
+@union
+@dataclass(frozen=True)
+class PublishFieldSet(Generic[_T], FieldSet, metaclass=ABCMeta):
+    """The fields necessary to publish an asset from a target.
+
+    Implementing rules should subclass this class ... and return a PublishPackageArtifactRequest as
+    output.
+    """
+
+    publish_request_type: ClassVar[Type[_T]]
+
+    @final
+    def request(self, packages: tuple[BuiltPackage, ...]) -> _T:
+        return self.publish_request_type(field_set=self, packages=packages)
+
+    @final
+    @classmethod
+    def rules(cls) -> tuple[UnionRule, ...]:
+        return (
+            UnionRule(PublishFieldSet, cls),
+            UnionRule(PublishPackagesRequest, cls.publish_request_type),
+        )
+
+
+@dataclass(frozen=True)
+class PublishPackageProcesses:
+    """Process to run in order to publish named artifact.
+
+    There are multiple names to support processes working on several packages for each process.
+    There are multiple processes to support publishing to several upstream services for each
+    package.
+    """
+
+    names: tuple[str, ...]
+    processes: tuple[InteractiveProcess, ...] = ()
+    description: str | None = None
+
+
+@dataclass(frozen=True)
+class PublishPackagesProcesses:
```

Comment:
These two... we almost have to come up with better naming here, so confusing right now..

Reply:
OK, I'm trying with `PublishPackages` as the lowest level operation, encapsulating a single process for one or more pacakges, and `PublishProcesses` is just a collection of the former.